Bugzilla/JobQueue/Runner.pm

```diff
@@ -173,42 +185,71 @@ sub gd_check {
     print get_text('job_queue_depth', { count => $count }) . "\n";
 }
 
+# override this to use IO::Async.
 sub gd_setup_signals {
     my $self = shift;
-    $self->SUPER::gd_setup_signals();
-    $SIG{TERM} = sub { $self->gd_quit_event(); }
+    my @signals = (
+        # Daemon::Generic by default handled these, though it had special treatment for HUP.
+        'INT',
+        'HUP',
+        # Bugzilla adds this
+        'TERM'
+    );
+    $self->{_signal_future} = Future->wait_any( map { catch_signal( $_, $_ ) } @signals );
 }
 
 sub gd_other_cmd {
     my ($self) = shift;
     if ($ARGV[0] eq "once") {
-        $self->_do_work("work_once");
-
-        exit(0);
+        Bugzilla::JobQueue::Worker->run("work_once");
+        exit;
     }
-    
+
     $self->SUPER::gd_other_cmd();
 }
 
+sub gd_quit_event { FATAL("gd_quit_event() should never be called") }
+sub gd_reconfig_event { FATAL("gd_reconfig_event() should never be called") }
+
 sub gd_run {
     my $self = shift;
 
-    $self->_do_work("work");
+    # This is so the process shows up in (h)top in a useful way.
+    local $PROGRAM_NAME = "$self->{gd_progname} [supervisor]";
+    my $code = $self->run_worker("work")->get;
+    unlink($self->{gd_pidfile});
+    exit $code;
 }
 
-sub _do_work {
+# This executes the script "jobqueue-worker.pl"
+# $EXECUTABLE_NAME is the name of the perl interpreter.
+sub run_worker {
     my ($self, $fn) = @_;
 
-    my $jq = Bugzilla->job_queue();
-    $jq->set_verbose($self->{debug});
-    foreach my $module (values %{ Bugzilla::JobQueue->job_map() }) {
-        eval "use $module";
-        $jq->can_do($module);
+    my $script = catfile( bz_locations->{cgi_path}, "jobqueue-worker.pl" );
```

Comment:
looks like perlcritic hasn't been run on this code - `ValuesAndExpressions::ProhibitInterpolationOfLiterals` would have caught those double quotes.

Reply:
It would have caught them, but because of progressive checking we don't usually care. So tests pass because there are exactly the same number of double quoted strings.

Reply:
It seems the vscode extension  doesn't show things with a severity below 2. I've filed #465 to make our severity higher for most of the ones that start at '1'.

I could raise the issue in https://github.com/sfodje/perlcritic but I think it's probably the right behavior.